### ISSUE TITLE
Get a list of opened windows

### ISSUE THREAD

**Issue by pranjlp** [ifreund/river#350]
Is there a command implemented to list currently opened windows and the currently focused window. I tried setting up a window switcher and couldn't find commands  like gettree in swaymsg or list-views etc.

**Comment by Leon-Plickat** [comment]
https://git.sr.ht/%7Eleon_plickat/lswt

**Comment by ifreund** [comment]
perhaps it makes sense to add such functionality to riverctl eventually for convenience, but I think keeping riverctl as simple as possible for the forseeable future is the right way to go and this functionality is already available from other tools.

**Comment by HeavyRain266** [comment]
Mayby consider adding list of tools (to wiki or readme) for managing outputs, getting lists of windows and other types to stop people from requesting those features in riverctl?

**Comment by novakne** [comment]
I mean there is already a [wiki page](https://github.com/ifreund/river/wiki/Recommended-Software) for this

**Comment by pranjlp** [comment]
Currently, you can focus a view only in a list fashion using focus-view... A window switcher needs a window ID to switch to it.. Is it possible to write a wm agnostic tool to set focus on vwindows in wayland like wmctrl?
Just leaving the question for discussion..

**Comment by Leon-Plickat** [comment]
@pranjlp I believe when you think of a window switcher, you think of some shell script. That will not work, precisely because river does not have the concept of assigning some ID to windows and allowing specific windows to be targeted by that ID. And even if river did, it would certainly not be compositor agnostic, because other compositors don't do that or do it differently. So just forget about scripting this.

But a compositor agnostic window switcher is perfectly possible. All you'll need is to write a Wayland client implementing the foreign-toplevel-management-v1 protocol. It allows clients to receive toplevel-handles, which represent windows, and ask the compositor to focus them, among other possibilities. I actually wanted to write something like that, I just haven't gotten around to it yet. The only problem is that currently river will ignore activation requests when the window is not visible on any output, but I am working on changing that eventually.

**Comment by JoshuaCrewe** [comment]
@Leon-Plickat 

>  All you'll need is to write a Wayland client implementing the foreign-toplevel-management-v1 protocol. 

Is this what https://git.sr.ht/~leon_plickat/lswt is doing? (Link taken from the wiki)

I am trying to replace wmctrl which has been making run-or-raise a possibility. Is this functionally possible in Wayland with river currently? I am at the start of my investigations.

**Comment by Leon-Plickat** [comment]
@JoshuaCrewe let me try again to explain it again, hopefully I can get my point across a bit clearer this time. First we'll look at a window switcher as an example, since you asked about it, then later I'll return to the case of run-or-raise (which is a lot simpler).

Imagine the following improvised window-switcher command: `list-windows | menu | focus-window`, where `list-windows` is a command that will list all windows currently opened, `menu` is some arbitrary menu program that allows you to select one of the entries, and `focus-window` is a command that will focus the window you selected.

Doing things over pipes means you are facing two problems:

### Problem 1: Missing atomicity.

From the perspective of river, that window switcher is actually two separate clients with seemingly no connection. In other words, controlling window management via shell script is inherently a bit hacky. It is hacky on X as well, but the average X user / developer apparently has a higher tolerance for hacks. However, for the two cases we'll look at here are this is not a problem, so we can ignore it here. I just mentioned it so we can get in the right head-space.

### Problem 2: How is the information "**this** exact window" passed through the pipes?

Answer: Not at all.

A Wayland client implementing the foreign toplevel extension can see all windows, however it uses the handle object(-id) it gets from the server to refer to windows. Wayland object ids can not only be recycled but (IIRC) vary for the client-object of different clients, even if they refer to the same server object. This means there is no token, no data, no id you can pass through the pipe that uniquely identifies this one exact window between the two Wayland clients in this window switcher example.

Some foreign toplevel implementing clients currently use the order in which windows are advertised as such an ID, but since the order is not guaranteed to be stable and since a window can be closed and therefore the order changed between `list-windows` and `select-windows`, this is not a particularly good things to do. It will probably work *most* of the times, but only by accident.

You can also neither use the app-id, because it is common to have multiple windows of the exact same app-id, nor the title, because you can not assume that the title of a window is constant.

### Doing it the Right Way™

There are two ways to correctly implement a window switcher. The first would be to add some per-window unique string token to the foreign toplevel protocol extension (or create a new extension for this). That however is a long-term thing and unlikely to happen soon(-ish), if at all.

The other way is to combine `list-windows` and `focus-window` into a single client. If you still want to have an external menu program, all you'd need to do is generate unique ids for each window client-side, pipe a list of them plus window titles to the menu and hope you'll get one of your unique ids back. Since this is a single client, the Wayland connection and therefore all the bound objects are kept between the list-windows and the focus-windows step, so whatever you get back from the menu program (assuming it's not a garbage value) will uniquely refer to one exact window in the current sessions history and should it still exist then you can map the ID back to a handle, since client-side you have exactly what the server is not exposing: an `ID<->window` map (which, as you by now can already guess, can not be transferred to other clients, based on the reasons outlined above). No false positives.

### run-or-raise

Luckily, the run-or-raise case is **a lot** more simple. Here you generally only care about app-ids of programs, so instead of worrying about identifying exact windows, just matching the first one with a matching app-id is enough. The Big Desktops do it this way as well, just that they tend to integrate everything into their desktop UI, while you - I assume - are interested in something more script-friendly.

All you'd really need is a program that tries to focus ("raise") a window with a user-specified app-id, and exits with either 0 or 1, based on whether it could find a window with the specified app-id. Then, assuming you know the command you want to run and the app-id of the window it will spawn:

```sh
focus-window --app-id="firefox" || rivrectl spawn firefox
```

In fact, such a program already exists: [wlrctl](https://git.sr.ht/~brocellous/wlrctl)



But there is a catch: River currently does not allow it to focus windows that are not already visible on an output.

I hope this made everything a bit clearer.

**Comment by JoshuaCrewe** [comment]
@Leon-Plickat thank you very much for taking the time to write all that out. I appreciate the effort you have put in for my benefit!

I think I am following what you are saying. I need to read over a few times.

It sounds like, for river, the bigger block is that you can't focus non visible output, that will hinder a run or raise workflow. Is non visible things on a different workspace/tag ?

**Comment by Leon-Plickat** [comment]
> It sounds like, for river, the bigger block is that you can't focus non visible output, that will hinder a run or raise workflow. Is non visible things on a different workspace/tag?

There is nothing magic about "hidden" windows. They are just windows that are currently not displayed, based on the tags of the window and the tags focused by the output.

The blocker is, that there are many ways river could respond to a "focus this window please" request as defined in the foreign toplevel protocol extension.

Should river...

1. ... move that window to the current tag(s)?
2. ... add the focused tag(s) to the window?
3. ... focus (only) the tag(s) of that window?
4. ... add the tag(s) of that window to the currently focused tags?

All of these will result in the window being displayed as the user wishes, but with varying side-effects. All of these are super trivial to implement and all of these do make a certain amount of sense.

The problem is, which one is the _right_ behaviour? That is not an easy decision, so it has been left undecided for now.

**Comment by sje30** [comment]
Thank you for this issue;   I am just starting to look at river (and wayland in general), and run-or-raise functionality is something I've had before in bspwm.

**Comment by pranjlp** [comment]
> The problem is, which one is the _right_ behaviour? That is not an easy decision, so it has been left undecided for now.
@Leon-Plickat  The blocker you speak of, could be left on the user. If river provides a string id and the tags of that window. The user can decide what he wants to do with it.
The key point here is that input focus is still a property bind to windows. visibility goes with tags.
The flexibility which river is trying to achieve could even workflows we haven't used till yet. Like a Window adder instead of a switcher as you mentioned in the 4th case. It only makes sense we leave it to the user. Maybe someone would want to use the zoom command on the window too to bring it to to the main area.
All river or the other client program needs to concern itself with is to provide all the relevant information and a method to change focus.
